core/src/main/scala/kafka/log/ProducerStateManager.scala

```diff
@@ -107,13 +107,18 @@ private[log] class ProducerAppendInfo(val producerId: Long,
         throw new DuplicateSequenceNumberException(s"Duplicate sequence number for producerId $producerId: (incomingBatch.firstSeq, " +
           s"incomingBatch.lastSeq): ($firstSeq, $lastSeq), (lastEntry.firstSeq, lastEntry.lastSeq): " +
           s"(${this.firstSeq}, ${this.lastSeq}).")
-      } else if (firstSeq != this.lastSeq + 1L) {
+      } else if (firstSeq != this.lastSeq + 1L && (firstSeq != 0 && this.lastSeq != Short.MaxValue)) {
         throw new OutOfOrderSequenceException(s"Out of order sequence number for producerId $producerId: $firstSeq " +
           s"(incoming seq. number), ${this.lastSeq} (current end sequence number)")
       }
     }
   }
 
+  private def isProducerFenced(producerEpoch: Short): Boolean = {
+      producerEpoch < 0 || (producerEpoch < this.producerEpoch &&
+        ((Short.MaxValue - this.producerEpoch) + producerEpoch > ProducerStateManager.ProducerEpochWraparoundDelta))
```

Comment:
We should log a message if we hit this case. The idea is that if we hit the awful situation of a partition being left behind, we will at least have a chance of being alerted to the problem.